internal/app/machined/pkg/runtime/v1alpha1/platform/metal/metal.go

```diff
@@ -64,28 +72,77 @@ func (m *Metal) Configuration(ctx context.Context) ([]byte, error) {
 }
 
 // PopulateURLParameters fills in empty parameters in the download URL.
-func PopulateURLParameters(downloadURL string, getSystemUUID func() (string, error)) (string, error) {
-	u, err := url.Parse(downloadURL)
+//nolint:gocyclo
+func PopulateURLParameters(ctx context.Context, downloadURL string, r state.State) (string, error) {
+	populatedURL := downloadURL
+
+	keyToVar := func(key string) string {
+		return `${` + key + `}`
+	}
+
+	genErr := func(varOfKey string, errToWrap error) error {
+		return fmt.Errorf("error while substituting %s: %w", varOfKey, errToWrap)
+	}
+
+	u, err := url.Parse(populatedURL)
 	if err != nil {
-		return "", fmt.Errorf("failed to parse %s: %w", constants.KernelParamConfig, err)
+		return "", fmt.Errorf("failed to parse %s: %w", populatedURL, err)
 	}
 
 	values := u.Query()
 
+	substitute := func(key string, getFunc func(ctx context.Context, r state.State) (string, error)) error {
+		for qKey, qValues := range values {
+			if len(qValues) > 0 {
+				qVal := qValues[0]
+
+				varOfKey := keyToVar(key)
+				varOfKeyExpr := regexp.QuoteMeta(varOfKey)
+				matcher := regexp.MustCompile(`(?i)` + varOfKeyExpr)
+
+				if matcher.MatchString(qVal) {
```

Comment:
similar to the above, probably it's easier to read if we `continue` if it doesn't match